Applications/Spire/Source/KeyBindings/KeyBindingsTableView.cpp

```diff
@@ -1,52 +1,97 @@
 #include "Spire/KeyBindings/KeyBindingsTableView.hpp"
 #include <QHeaderView>
 #include <QLabel>
+#include <QMouseEvent>
 #include <QPainter>
 #include <QScrollBar>
 #include <QVBoxLayout>
 #include "Spire/KeyBindings/KeySequenceItemDelegate.hpp"
 #include "Spire/Spire/Dimensions.hpp"
+#include "Spire/Spire/Utility.hpp"
+#include "Spire/Ui/IconButton.hpp"
 
 using namespace Spire;
 
+namespace {
+  auto create_delete_button(QWidget* parent) {
+    auto button_size = scale(16, 16);
+    auto close_box = QRect(QPoint(4, 4), scale(8, 8));
+    return new IconButton(
+      imageFromSvg(":/Icons/close-purple.svg", button_size, close_box),
+      imageFromSvg(":/Icons/close-red.svg", button_size, close_box), parent);
+  }
+
+  auto DELETE_ROW_LAYOUT_WIDTH() {
+    static auto width = scale_width(26);
+    return width;
+  }
+
+  auto HEADER_PADDING() {
+    static auto padding = scale_width(17);
+    return padding;
+  }
+
+  auto TABLE_PADDING() {
+    static auto padding = scale_width(8);
+    return padding;
+  }
+
+  auto bounded_value(int value, int max) {
+    if(value < 0) {
+      return max;
+    } else if(value > max) {
+      return 0;
+    }
+    return value;
+  }
+}
+
 KeyBindingsTableView::KeyBindingsTableView(QHeaderView* header,
-    QWidget* parent)
+    bool can_delete_rows, QWidget* parent)
     : ScrollArea(true, parent),
-      m_header(header) {
+      m_header(header),
+      m_can_delete_rows(can_delete_rows),
+      m_is_default_cell_selected(true),
+      m_is_editing_cell(false) {
   m_header->setParent(this);
+  m_header->setStretchLastSection(false);
   auto main_widget = new QWidget(this);
-  auto layout = new QVBoxLayout(main_widget);
-  layout->setContentsMargins({});
-  layout->setSpacing(0);
   connect(m_header, &QHeaderView::sectionResized, this,
     &KeyBindingsTableView::on_header_resize);
   connect(m_header, &QHeaderView::sectionMoved, this,
     &KeyBindingsTableView::on_header_move);
-  auto header_padding = new QWidget(this);
-  header_padding->setFixedHeight(m_header->height());
-  layout->addWidget(header_padding);
-  m_table = new CustomGridTableView(this);
-  layout->addWidget(m_table);
+  m_table = new CustomGridTableView(main_widget);
```

Comment:
Can you expand on this? Do you mean assign the new table to a temporary then assign that temporary to m_table?

Reply:
Yes. It is a general principle: prefer atomic updates to the program state. A temporary table would bear less state to the program than a member of the View, so it is preferred if all modifications to it are done before assignment.
You can also avoid creation of a named temporary variable using an immediately called lambda.

Reply:
The majority of the objects in the view and model classes for Spire are constructed/assigned, then modified. I can see in other cases where it's a good idea to use a temporary, but for a single-threaded GUI I'm not sure that it's worth it to go back and modify all of the other cases that are like this.

Reply:
That principle has little to do with threads and whether the application uses GUI or not. It might deal with exception safety and such under certain conditions, but the primary consideration for us is standardization.

Now, the rationale for code standardization is not all about improving the quality of a software product itself. The rules may be based on software quality concerns (among others, such as code readability, development speed, etc.), yet the main idea behind any kind of a standard is that there is (ideally) only one proper way of doing a thing. Using the alphabetic order for includes is not going to improve the quality of a software product, yet if we had different beliefs on the correct order we could end up altering each other's code every time we modify the file.

Then, when something is brought up during a code review, you should not apply similar changes to the existing code unless it is a critical bug. If we figure out that the existing code needs to be modified, a separate change should be mad specifically for that purpose so that the code for the feature is not messed up. It does not hurt, however, to incorporate a practice into the code added by the current change or to keep it in mind while developing new features.

Back to the issue brought up here, I do not think it is something we are really strict about. It is more of a fundamental principle in software development (minimizing intermediate state) than a coding style concern. I did not stress it (though I brought it up at some point) before because there were more important issues, while I do not see many problems in this PR, especially for a change of this size. That said, I think it is a good idea to apply the principle in your future changes, though you can keep the code in this PR as is if you find it difficult to navigate or have better uses for your time.